### PR TITLE
React refactorings

### PR DESCRIPTION
**Pull request by brainkim**
This PR eliminates the XData class-based abstractions which were used to implement the React hooks, in favor of hooks which are more “idiomatic” to React. The old class-based abstractions generated an immense amount of mental overhead, insofar as it attempted to abstract over React hooks by defining two methods `execute()` and `afterExecute()` which are called by the hook body and in a `useEffect()` hook respectively.

This PR should make the following issues easier to fix:
- [x] #5140 
- [x] #5531
- [x] #6636
- [x] #6816
- [x] #7171
- [x] #7243
- [x] #7388
- [x] #7396
- [x] #7484
- [x] #7543
- [x] #7564
- [x] #8270
- [x] #8350
- [x] #8313

### CODE REVIEW

File: src/react/hooks/useQuery.ts [hunk 1]
```diff
@@ -1,16 +1,355 @@
...
+      options?.ssr !== false &&
+      !options?.skip &&
+      obsQuery.getCurrentResult().loading
+    ) {
+      // TODO: This is a legacy API which could probably be cleaned up
+      context.renderPromises.addQueryPromise(
+        {
+          // The only options which seem to actually be used by the
+          // RenderPromises class are query and variables.
+          getOptions: () => createWatchQueryOptions(query, options),
```
Comment (benjamn) [inline src/react/hooks/useQuery.ts:64]:
```suggestion
          getOptions: () => watchQueryOptions,
```
Do we need to recreate these options?

File: src/react/hooks/useQuery.ts [hunk 2]
```diff
@@ -268,7 +268,7 @@ export function useQuery<
       error: void 0,
       networkStatus: NetworkStatus.loading,
     };
-  } else if (options?.skip) {
+  } else if (options?.skip || options?.fetchPolicy === 'standby') {
```
Comment (benjamn) [inline src/react/hooks/useQuery.ts:271]: Glad to see this! 👍

File: src/react/hooks/useQuery.ts [hunk 3]
```diff
@@ -107,26 +107,26 @@ export function useQuery<
...
     result,
     data: void 0 as TData | undefined,
+    watchQueryOptions: createWatchQueryOptions(query, options),
   });
 
   // An effect to recreate the obsQuery whenever the client or query changes.
   // This effect is also responsible for checking and updating the obsQuery
   // options whenever they change.
   useEffect(() => {
+    const watchQueryOptions = createWatchQueryOptions(query, options);
```
Comment (benjamn) [inline src/react/hooks/useQuery.ts:117]: Why recreate these options, instead of using `ref.current.watchQueryOptions`?

File: src/core/ObservableQuery.ts [hunk 4]
```diff
@@ -288,7 +288,9 @@ export class ObservableQuery<
     // (no-cache, network-only, or cache-and-network), override it with
     // network-only to force the refetch for this fetchQuery call.
     const { fetchPolicy } = this.options;
-    if (fetchPolicy === 'no-cache') {
+    if (fetchPolicy === 'standby') {
+      // do nothing
```
Comment (benjamn) [inline src/core/ObservableQuery.ts:292]: I think this has the effect of making `reobserveOptions.fetchPolicy` default to `cache-first`. Is that really what we want when `standby` queries are refetched?

File: src/react/hooks/useLazyQuery.ts [hunk 5]
```diff
@@ -43,6 +43,16 @@ export function useLazyQuery<TData = any, TVariables = OperationVariables>(
...
   if (!execution.called) {
+    for (const key in result) {
+      if (typeof (result as any)[key] === 'function') {
+        const method = (result as any)[key];
+        (result as any)[key] = (...args: any) => {
+          setExecution({ called: true });
+          return method(...args);
+        };
+      }
+    }
```
Comment (benjamn) [inline src/react/hooks/useLazyQuery.ts:63]: This makes me nervous. Can we do something less dynamic here, based on known method properties?

File: src/react/hooks/useQuery.ts [hunk 6]
```diff
@@ -1,16 +1,355 @@
...
+  });
+
+  let [result, setResult] = useState(() => {
+    const result = obsQuery.getCurrentResult();
+    if (!result.loading) {
+      if (result.data) {
+        options?.onCompleted?.(result.data);
+      } else if (result.error) {
+        options?.onError?.(result.error);
+      }
```
Comment (benjamn) [inline src/react/hooks/useQuery.ts:98]:
```suggestion
    if (!result.loading && options) {
      if (result.error) {
        options.onError?.(result.error);
      } else if (result.data) {
        options.onCompleted?.(result.data);
      }
```
If I'm reading the [`onCompleted` and `onError` docs](https://www.apollographql.com/docs/react/data/queries/#oncompleted) correctly, we should give `result.error` preference here, and in the couple of other places where we call `onCompleted` and `onError`.

File: src/react/hooks/useSubscription.ts [hunk 7]
```diff
@@ -1,56 +1,122 @@
...
+      query: subscription,
+      variables: options?.variables,
+      fetchPolicy: options?.fetchPolicy,
+      context: options?.context,
+    });
+  });
 
-  if (__DEV__) {
-    // ensure we run an update after refreshing so that we can resubscribe
-    useAfterFastRefresh(forceUpdate);
```
Comment (benjamn) [inline src/react/hooks/useSubscription.ts:44]: Are we losing this functionality, or is it no longer needed after the refactoring?

**Review by benjamn — COMMENTED** [review]
This is really exciting!

Could you rebase against the `release-3.5` branch, so we can put out a beta release with these changes?

**Review by hwillson — APPROVED** [review]
Really incredible work @brainkim! I don't have much to add onto Ben's comments other than a reminder to update the `CHANGELOG`. Thanks for all of this! 🎉

File: src/core/ObservableQuery.ts (hunk 4 above)
Comment (brainkim) [reply src/core/ObservableQuery.ts:292]: The `reobserve()` method will merge the observable query options with the options passed in, so the effect of falling through and not assigning a `fetchPolicy` property in the `refetch()` method will have the effect of using the original `fetchPolicy`. Non-ideal in terms of readability, as you noticed. I’m happy to be more explicit though.

Comment (brainkim) [reply src/core/ObservableQuery.ts:292]: The behavior is also tested in this unit test. https://github.com/apollographql/apollo-client/blob/d4e4799b0004e23afb471e2b2050483ed1f2079a/src/react/hooks/__tests__/useQuery.test.tsx#L2723-L2756

File: src/react/hooks/useSubscription.ts [hunk 8]
```diff
@@ -1,56 +1,122 @@
...
+    }
+
+    if (options?.skip && !options?.skip !== !ref.current.options?.skip) {
+      setResult({
+        loading: false,
+        data: void 0,
+        error: void 0,
+        variables: options?.variables,
+      });
+      setObservable(null);
```
[the remark this answers is not in the document]
Comment (brainkim) [reply src/react/hooks/useSubscription.ts:54]: I’m not sure, but my guess is no. There are a lot of subtle timing changes which this PR inevitably produces as we refactor state calls.

File: src/react/hooks/useQuery.ts (hunk 3 above)
Comment (brainkim) [reply src/react/hooks/useQuery.ts:117]: This effect is also the effect which sets `ref.current.watchQueryOptions()`. That function runs on the scale of microseconds so I’d rather have fresh `watchQueryOptions` rather than reasoning about using it from a React hook closure.

File: src/react/hooks/useLazyQuery.ts (hunk 5 above)
Comment (brainkim) [reply src/react/hooks/useLazyQuery.ts]: Done. I’m thinking all methods should execute `useLazyQuery` except for `stopPolling()` does that make sense?

File: src/react/components/__tests__/client/Mutation.test.tsx [hunk 9]
```diff
@@ -922,7 +926,7 @@ describe('General Mutation testing', () => {
     });
   });
 
-  it('allows a refetchQueries prop as string and variables have updated', async () => {
+  it('allows a refetchQueries prop as string and variables have updated', async () => new Promise((resolve, reject) => {
```
[the remark this answers is not in the document]
Comment (brainkim) [reply src/react/components/__tests__/client/Mutation.test.tsx:934]: Fine!

File: src/react/hooks/useQuery.ts (hunk 1 above)
Comment (brainkim) [reply src/react/hooks/useQuery.ts:64]: `getOptions()` is called in a `useState` callback, and right now it only seems to have the initial hook options. I can’t reference the ref which holds a copy of options and such because it doesn’t exist yet. It’s a conundrum.

File: src/react/hooks/useQuery.ts (hunk 6 above)
Comment (brainkim) [reply src/react/hooks/useQuery.ts]: Done. I think it might be nice to streamline the result modification and callback calling code somehow, but haven’t figured out where to do it. For instance, we call onError in the observable onError callback and onCompleted in the observable onNext callbacks, but should these methods be called someplace else?

File: src/react/hooks/useSubscription.ts (hunk 7 above)
Comment (brainkim) [reply src/react/hooks/useSubscription.ts:44]: I’m not sure; I’ll have to look into it. We don’t call side-effects from the render function for the most part anymore, except for certain edge cases where I could not get the tests working.

File: src/react/hooks/useQuery.ts (hunk 6 above)
Comment (benjamn) [reply src/react/hooks/useQuery.ts]: Do we have a `Promise<ApolloQueryResult<TData>>` somewhere that we could attach `onCompleted` and/or `onError` to, as `then` handlers?

Comment (brainkim) [reply src/react/hooks/useQuery.ts]: Nope. The difficult part is that an “ApolloQueryResult” is created in multiple places, in the initial useState callback, in the observable callbacks, and based on whether or not the query is skipped. Trying to abstract over all these cases is hard, especially because any function we define inline might close over stale variables.

File: src/react/hooks/useSubscription.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (benjamn) [reply src/react/hooks/useSubscription.ts:54]: @brainkim Maybe let's use `unstable_batchedUpdates` then, just in case? Or are you saying it's important to preserve the timing of the separate updates?

File: src/react/hooks/useQuery.ts (hunk 1 above)
Comment (benjamn) [reply src/react/hooks/useQuery.ts:64]: Okay, good point about this code being SSR-specific. Suggestion withdrawn! 😄

File: src/react/hooks/useSubscription.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (brainkim) [reply src/react/hooks/useSubscription.ts:54]: @benjamn 
It is impossible to preserve the exact timings of the old React hooks implementation, insofar as it used a `useReducer/forceUpdate` pattern, which, among other things, would in many circumstances, call setState directly in the render function. And I’m still not sure that “batching” the two state updates would matter, given that the result and observable are two independently updating pieces of state.

I’m not familiar with `unstable_batchedUpdates`. Have you adopted it somewhere in this codebase?
- What is availability of `unstable_batchedUpdates`?
- Is it okay that it seems to be defined on `ReactDOM`, when none of the hooks require the `ReactDOM` package, especially given that some of our poor users are using React Native?
- Is it a merely an optimization?
- Would it hamper preact support?

**Comment by brainkim** [comment]
@benjamn Can I get a review for [this commit](https://github.com/apollographql/apollo-client/pull/8596/commits/cf90a3d8c67a953888e984a9ee6dc0a0d7e884ff) which would solve https://github.com/apollographql/apollo-client/issues/8137
